File: android/src/main/kotlin/com/usercentrics/sdk/flutter/serializer/CMPDataSerializer.kt
package com.usercentrics.sdk.flutter.serializer

import com.usercentrics.sdk.UsercentricsCMPData
import com.usercentrics.sdk.v2.location.data.UsercentricsLocation
import com.usercentrics.sdk.v2.settings.data.CCPASettings
import com.usercentrics.sdk.v2.settings.data.CustomizationColor
import com.usercentrics.sdk.v2.settings.data.CustomizationFont
import com.usercentrics.sdk.v2.settings.data.FirstLayer
import com.usercentrics.sdk.v2.settings.data.PublishedApp
import com.usercentrics.sdk.v2.settings.data.SecondLayer
import com.usercentrics.sdk.v2.settings.data.TCF2ChangedPurposes
import com.usercentrics.sdk.v2.settings.data.TCF2Settings
import com.usercentrics.sdk.v2.settings.data.UsercentricsCategory
import com.usercentrics.sdk.v2.settings.data.UsercentricsCustomization
import com.usercentrics.sdk.v2.settings.data.UsercentricsLabels
import com.usercentrics.sdk.v2.settings.data.UsercentricsService
import com.usercentrics.sdk.v2.settings.data.UsercentricsSettings
import com.usercentrics.sdk.v2.settings.data.VariantsSettings
import com.usercentrics.sdk.v2.translation.data.LegalBasisLocalization
import com.usercentrics.sdk.v2.translation.data.TranslationAriaLabels

internal fun UsercentricsCMPData.serialize(): Any {
    return mapOf(
        "settings" to settings.serialize(),
        "services" to services.map { it.serialize() },
        "categories" to categories.map { it.serialize() },
        "activeVariant" to activeVariant.name,
        "userLocation" to userLocation.serialize(),
        "legalBasis" to legalBasis.serialize()
    )
}

private fun UsercentricsSettings.serialize(): Any {
    return mapOf(
        "labels" to labels.serialize(),
        "showInitialViewForVersionChange" to showInitialViewForVersionChange,
        "reshowBanner" to reshowBanner,
        "displayOnlyForEU" to displayOnlyForEU,
        "secondLayer" to secondLayer.serialize(),
        "cookiePolicyUrl" to cookiePolicyUrl,
        "tcf2" to tcf2?.serialize(),
        "ccpa" to ccpa?.serialize(),
        "privacyPolicyUrl" to privacyPolicyUrl,
        "firstLayer" to firstLayer?.serialize(),
        "imprintUrl" to imprintUrl,
        "firstLayerDescriptionHtml" to firstLayerDescriptionHtml,
        "bannerMobileDescriptionIsActive" to bannerMobileDescriptionIsActive,
        "firstLayerMobileDescriptionHtml" to firstLayerMobileDescriptionHtml,
        "version" to version,
        "language" to language,
        "tcf2Enabled" to tcf2Enabled,
        "settingsId" to settingsId,
        "languagesAvailable" to languagesAvailable,
        "enablePoweredBy" to enablePoweredBy,
        "editableLanguages" to editableLanguages,
        "customization" to customization?.serialize(),
        "variants" to variants?.serialize(),
        "dpsDisplayFormat" to (dpsDisplayFormat?.name ?: ""),
        "framework" to (framework?.name ?: ""),
        "publishedApps" to publishedApps?.map { it.serialize() },
        "renewConsentsTimestamp" to renewConsentsTimestamp,
        "consentWebhook" to consentWebhook
    )
}

private fun UsercentricsLabels.serialize(): Any {
    return mapOf(
        "btnAcceptAll" to btnAcceptAll,
        "btnDeny" to btnDeny,
        "btnSave" to btnSave,
        "firstLayerTitle" to firstLayerTitle,
        "accepted" to accepted,
        "denied" to denied,
        "date" to date,
        "decision" to decision,
        "dataCollectedList" to dataCollectedList,
        "dataCollectedInfo" to dataCollectedInfo,
        "locationOfProcessing" to locationOfProcessing,
        "transferToThirdCountries" to transferToThirdCountries,
        "dataPurposes" to dataPurposes,
        "dataPurposesInfo" to dataPurposesInfo,
        "dataRecipientsList" to dataRecipientsList,
        "descriptionOfService" to descriptionOfService,
        "history" to history,
        "historyDescription" to historyDescription,
        "legalBasisList" to legalBasisList,
        "legalBasisInfo" to legalBasisInfo,
        "processingCompanyTitle" to processingCompanyTitle,
        "technologiesUsed" to technologiesUsed,
        "technologiesUsedInfo" to technologiesUsedInfo,
        "cookiePolicyInfo" to cookiePolicyInfo,
        "optOut" to optOut,
        "policyOf" to policyOf,
        "imprintLinkText" to imprintLinkText,
        "privacyPolicyLinkText" to privacyPolicyLinkText,
        "categories" to categories,
        "anyDomain" to anyDomain,
        "day" to day,
        "days" to days,
        "domain" to domain,
        "duration" to duration,
        "informationLoadingNotPossible" to informationLoadingNotPossible,
        "hour" to hour,
        "hours" to hours,
        "identifier" to identifier,
        "maximumAgeCookieStorage" to maximumAgeCookieStorage,
        "minute" to minute,
        "minutes" to minutes,
        "month" to month,
        "months" to months,
        "multipleDomains" to multipleDomains,
        "no" to no,
        "nonCookieStorage" to nonCookieStorage,
        "seconds" to seconds,
        "session" to session,
        "loadingStorageInformation" to loadingStorageInformation,
        "storageInformation" to storageInformation,
        "detailedStorageInformation" to detailedStorageInformation,
        "tryAgain" to tryAgain,
        "type" to type,
        "year" to year,
        "years" to years,
        "yes" to yes,
        "storageInformationDescription" to storageInformationDescription,
        "btnBannerReadMore" to btnBannerReadMore,
        "linkToDpaInfo" to linkToDpaInfo,
        "second" to second,
        "consent" to consent,
        "secondLayerTitle" to secondLayerTitle,
        "secondLayerDescriptionHtml" to secondLayerDescriptionHtml,
        "btnMore" to btnMore,
        "retentionPeriod" to retentionPeriod,
        "explicit" to explicit,
        "transferToThirdCountriesInfo" to transferToThirdCountriesInfo,
        "more" to more,
        "headerModal" to headerModal,
        "furtherInformationOptOut" to furtherInformationOptOut,
        "cookiePolicyLinkText" to cookiePolicyLinkText,
        "noImplicit" to noImplicit,
        "yesImplicit" to yesImplicit,
        "readLess" to readLess,
    )
}

private fun CCPASettings.serialize(): Any {
    return mapOf(
        "optOutNoticeLabel" to optOutNoticeLabel,
        "btnSave" to btnSave,
        "firstLayerTitle" to firstLayerTitle,
        "isActive" to isActive,
        "showOnPageLoad" to showOnPageLoad,
        "reshowAfterDays" to reshowAfterDays,
        "iabAgreementExists" to iabAgreementExists,
        "appFirstLayerDescription" to appFirstLayerDescription,
        "firstLayerMobileDescriptionIsActive" to firstLayerMobileDescriptionIsActive,
        "firstLayerMobileDescription" to firstLayerMobileDescription,
        "secondLayerTitle" to secondLayerTitle,
        "secondLayerDescription" to secondLayerDescription,
        "secondLayerHideLanguageSwitch" to secondLayerHideLanguageSwitch,
        "btnMoreInfo" to btnMoreInfo,
    )
}

private fun TCF2Settings.serialize(): Any {
    return mapOf(
        "firstLayerTitle" to firstLayerTitle,
        "secondLayerTitle" to secondLayerTitle,
        "tabsPurposeLabel" to tabsPurposeLabel,
        "tabsVendorsLabel" to tabsVendorsLabel,
        "labelsFeatures" to labelsFeatures,
        "labelsIabVendors" to labelsIabVendors,
        "labelsNonIabPurposes" to labelsNonIabPurposes,
        "labelsNonIabVendors" to labelsNonIabVendors,
        "labelsPurposes" to labelsPurposes,
        "vendorFeatures" to vendorFeatures,
        "vendorLegitimateInterestPurposes" to vendorLegitimateInterestPurposes,
        "vendorPurpose" to vendorPurpose,
        "vendorSpecialFeatures" to vendorSpecialFeatures,
        "vendorSpecialPurposes" to vendorSpecialPurposes,
        "togglesConsentToggleLabel" to togglesConsentToggleLabel,
        "togglesLegIntToggleLabel" to togglesLegIntToggleLabel,
        "buttonsAcceptAllLabel" to buttonsAcceptAllLabel,
        "buttonsDenyAllLabel" to buttonsDenyAllLabel,
        "buttonsSaveLabel" to buttonsSaveLabel,
        "linksManageSettingsLabel" to linksManageSettingsLabel,
        "linksVendorListLinkLabel" to linksVendorListLinkLabel,
        "cmpId" to cmpId,
        "cmpVersion" to cmpVersion,
        "categoriesOfDataLabel" to categoriesOfDataLabel,
        "dataRetentionPeriodLabel" to dataRetentionPeriodLabel,
        "legitimateInterestLabel" to legitimateInterestLabel,
        "version" to version,
        "examplesLabel" to examplesLabel,
        "firstLayerHideToggles" to firstLayerHideToggles,
        "secondLayerHideToggles" to secondLayerHideToggles,
        "hideLegitimateInterestToggles" to hideLegitimateInterestToggles,
        "firstLayerHideButtonDeny" to firstLayerHideButtonDeny,
        "secondLayerHideButtonDeny" to secondLayerHideButtonDeny,
        "publisherCountryCode" to publisherCountryCode,
        "purposeOneTreatment" to purposeOneTreatment,
        "selectedVendorIds" to selectedVendorIds,
        "gdprApplies" to gdprApplies,
        "selectedStacks" to selectedStacks,
        "disabledSpecialFeatures" to disabledSpecialFeatures,
        "firstLayerShowDescriptions" to firstLayerShowDescriptions,
        "hideNonIabOnFirstLayer" to hideNonIabOnFirstLayer,
        "resurfacePeriodEnded" to resurfacePeriodEnded,
        "resurfacePurposeChanged" to resurfacePurposeChanged,
        "resurfaceVendorAdded" to resurfaceVendorAdded,
        "firstLayerDescription" to firstLayerDescription,
        "firstLayerAdditionalInfo" to firstLayerAdditionalInfo,
        "secondLayerDescription" to secondLayerDescription,
        "togglesSpecialFeaturesToggleOn" to togglesSpecialFeaturesToggleOn,
        "togglesSpecialFeaturesToggleOff" to togglesSpecialFeaturesToggleOff,
        "appLayerNoteResurface" to appLayerNoteResurface,
        "firstLayerNoteResurface" to firstLayerNoteResurface,
        "firstLayerMobileVariant" to (firstLayerMobileVariant?.name ?: ""),
        "showDataSharedOutsideEUText" to showDataSharedOutsideEUText,
        "dataSharedOutsideEUText" to dataSharedOutsideEUText,
        "vendorIdsOutsideEUList" to vendorIdsOutsideEUList,
        "scope" to scope.name,
        "changedPurposes" to changedPurposes?.serialize(),
        "acmV2Enabled" to acmV2Enabled,
        "selectedATPIds" to selectedATPIds
    )
}

private fun UsercentricsCustomization.serialize(): Any {
    return mapOf(
        "color" to color?.serialize(),
        "font" to font?.serialize(),
        "logoUrl" to logoUrl,
        "borderRadiusLayer" to borderRadiusLayer,
        "borderRadiusButton" to borderRadiusButton,
        "overlayOpacity" to overlayOpacity,
    )
}

private fun CustomizationFont.serialize(): Any {
    return mapOf(
        "family" to family,
        "size" to size,
    )
}

private fun CustomizationColor.serialize(): Any {
    return mapOf(
        "primary" to primary,
        "acceptBtnText" to acceptBtnText,
        "acceptBtnBackground" to acceptBtnBackground,
        "denyBtnText" to denyBtnText,
        "denyBtnBackground" to denyBtnBackground,
        "saveBtnText" to saveBtnText,
        "saveBtnBackground" to saveBtnBackground,
        "linkIcon" to linkIcon,
        "linkFont" to linkFont,
        "text" to text,
        "layerBackground" to layerBackground,
        "overlay" to overlay,
        "toggleInactiveBackground" to toggleInactiveBackground,
        "toggleInactiveIcon" to toggleInactiveIcon,
        "toggleActiveBackground" to toggleActiveBackground,
        "toggleActiveIcon" to toggleActiveIcon,
        "toggleDisabledBackground" to toggleDisabledBackground,
        "toggleDisabledIcon" to toggleDisabledIcon,
        "secondLayerTab" to secondLayerTab,
    )
}

private fun FirstLayer.serialize(): Any {
    return mapOf(
        "hideButtonDeny" to hideButtonDeny,
    )
}

private fun SecondLayer.serialize(): Any {
    return mapOf(
        "tabsCategoriesLabel" to tabsCategoriesLabel,
        "tabsServicesLabel" to tabsServicesLabel,
        "acceptButtonText" to acceptButtonText,
        "denyButtonText" to denyButtonText,
        "hideButtonDeny" to hideButtonDeny,
        "hideLanguageSwitch" to hideLanguageSwitch,
        "hideTogglesForServices" to hideTogglesForServices,
        "hideDataProcessingServices" to hideDataProcessingServices,
    )
}

private fun UsercentricsService.serialize(): Any {
    return mapOf(
        "templateId" to templateId,
        "version" to version,
        "type" to type,
        "dataProcessor" to dataProcessor,
        "dataPurposes" to dataPurposes,
        "processingCompany" to processingCompany,
        "nameOfProcessingCompany" to nameOfProcessingCompany,
        "addressOfProcessingCompany" to addressOfProcessingCompany,
        "descriptionOfService" to descriptionOfService,
        "languagesAvailable" to languagesAvailable,
        "dataCollectedList" to dataCollectedList,
        "dataPurposesList" to dataPurposesList,
        "dataRecipientsList" to dataRecipientsList,
        "legalBasisList" to legalBasisList,
        "retentionPeriodList" to retentionPeriodList,
        "subConsents" to subConsents,
        "language" to language,
        "linkToDpa" to linkToDpa,
        "legalGround" to legalGround,
        "optOutUrl" to optOutUrl,
        "policyOfProcessorUrl" to policyOfProcessorUrl,
        "categorySlug" to categorySlug,
        "retentionPeriodDescription" to retentionPeriodDescription,
        "dataProtectionOfficer" to dataProtectionOfficer,
        "privacyPolicyURL" to privacyPolicyURL,
        "cookiePolicyURL" to cookiePolicyURL,
        "locationOfProcessing" to locationOfProcessing,
        "dataCollectedDescription" to dataCollectedDescription,
        "thirdCountryTransfer" to thirdCountryTransfer,
        "description" to description,
        "cookieMaxAgeSeconds" to cookieMaxAgeSeconds,
        "usesNonCookieAccess" to usesNonCookieAccess,
        "deviceStorageDisclosureUrl" to deviceStorageDisclosureUrl,
        "isDeactivated" to isDeactivated,
        "disableLegalBasis" to disableLegalBasis,
        "isEssential" to isEssential,
        "technologyUsed" to technologyUsed,
        "deviceStorage" to deviceStorage.serializer(),
        "isHidden" to isHidden,
    )
}

private fun UsercentricsCategory.serialize(): Any {
    return mapOf(
        "categorySlug" to categorySlug,
        "label" to label,
        "description" to description,
        "isEssential" to isEssential
    )
}

internal fun UsercentricsLocation.serialize(): Any {
    return mapOf(
        "countryCode" to countryCode,
        "regionCode" to regionCode,
        "isInEU" to isInEU(),
        "isInUS" to isInUS(),
        "isInCalifornia" to isInCalifornia(),
    )
}

internal fun LegalBasisLocalization.serialize(): Any {
    return mapOf(
        "labelsAria" to labelsAria?.serialize(),
        "data" to data
    )
}

internal fun TranslationAriaLabels.serialize(): Any {
    return mapOf(
        "acceptAllButton" to acceptAllButton,
        "ccpaButton" to ccpaButton,
        "ccpaMoreInformation" to ccpaMoreInformation,
        "closeButton" to closeButton,
        "collapse" to collapse,
        "cookiePolicyButton" to cookiePolicyButton,
        "copyControllerId" to copyControllerId,
        "denyAllButton" to denyAllButton,
        "expand" to expand,
        "fullscreenButton" to fullscreenButton,
        "imprintButton" to imprintButton,
        "languageSelector" to languageSelector,
        "privacyButton" to privacyButton,
        "privacyPolicyButton" to privacyPolicyButton,
        "saveButton" to saveButton,
        "serviceInCategoryDetails" to serviceInCategoryDetails,
        "servicesInCategory" to servicesInCategory,
        "tabButton" to tabButton,
        "usercentricsCMPButtons" to usercentricsCMPButtons,
        "usercentricsCMPContent" to usercentricsCMPContent,
        "usercentricsCMPHeader" to usercentricsCMPHeader,
        "usercentricsCMPUI" to usercentricsCMPUI,
        "usercentricsCard" to usercentricsCard,
        "usercentricsList" to usercentricsList,
        "vendorConsentToggle" to vendorConsentToggle,
        "vendorDetailedStorageInformation" to vendorDetailedStorageInformation,
        "vendorLegIntToggle" to vendorLegIntToggle
    )
}

private fun VariantsSettings?.serialize(): Any? {
    if (this == null) {
        return null
    }
    return mapOf(
        "enabled" to enabled,
        "experimentsJson" to experimentsJson,
        "activateWith" to activateWith,
    )
}

private fun PublishedApp.serialize(): Any {
    return mapOf(
        "bundleId" to bundleId,
        "platform" to platform.name,
    )
}

private fun TCF2ChangedPurposes?.serialize(): Any? {
    if (this == null) {
        return null
    }
    return mapOf(
        "purposes" to purposes,
        "legIntPurposes" to legIntPurposes
    )
}

File: ios/usercentrics_sdk/Sources/usercentrics_sdk/Serializer/CMPDataSerializer.swift
import Usercentrics
import Flutter

extension UsercentricsCMPData {
    func serialize() -> Any {
        return [
            "settings": self.settings.serialize(),
            "services": self.services.map { $0.serialize() },
            "categories": self.categories.map { $0.serialize() },
            "activeVariant": self.activeVariant.name as Any,
            "userLocation": self.userLocation.serialize(),
            "legalBasis": self.legalBasis.serialize(),
        ]
    }
}

extension UsercentricsSettings {
    func serialize() -> Any {
        return [
            "labels" : self.labels.serialize(),
            "showInitialViewForVersionChange" : self.showInitialViewForVersionChange,
            "reshowBanner" : self.reshowBanner as Any,
            "displayOnlyForEU" : self.displayOnlyForEU,
            "secondLayer" : self.secondLayer.serialize() as Any,
            "cookiePolicyUrl" : self.cookiePolicyUrl as Any,
            "tcf2" : self.tcf2?.serialize() as Any,
            "ccpa" : self.ccpa?.serialize() as Any,
            "privacyPolicyUrl" : self.privacyPolicyUrl as Any,
            "firstLayer" : self.firstLayer?.serialize() as Any,
            "imprintUrl" : self.imprintUrl as Any,
            "firstLayerDescriptionHtml" : self.firstLayerDescriptionHtml as Any,
            "bannerMobileDescriptionIsActive" : self.bannerMobileDescriptionIsActive,
            "firstLayerMobileDescriptionHtml" : self.firstLayerMobileDescriptionHtml as Any,
            "version" : self.version,
            "language" : self.language,
            "tcf2Enabled" : self.tcf2Enabled,
            "settingsId" : self.settingsId,
            "languagesAvailable" : self.languagesAvailable,
            "enablePoweredBy" : self.enablePoweredBy,
            "editableLanguages" : self.editableLanguages,
            "customization" : self.customization?.serialize() as Any,
            "variants": (self.variants?.serialize() ?? nil) as Any,
            "dpsDisplayFormat": (self.dpsDisplayFormat?.name ?? "") as Any,
            "framework": (self.framework?.name ?? "") as Any,
            "publishedApps": (self.publishedApps?.map { $0.serialize() } ?? nil) as Any,
            "renewConsentsTimestamp": self.renewConsentsTimestamp as Any
        ]
    }
}

extension UsercentricsLabels {
    func serialize() -> Any {
        return [
            "btnAcceptAll" : self.btnAcceptAll,
            "btnDeny" : self.btnDeny,
            "btnSave" : self.btnSave,
            "firstLayerTitle" : self.firstLayerTitle,
            "accepted" : self.accepted,
            "denied" : self.denied,
            "date" : self.date,
            "decision" : self.decision,
            "dataCollectedList" : self.dataCollectedList,
            "dataCollectedInfo" : self.dataCollectedInfo,
            "locationOfProcessing" : self.locationOfProcessing,
            "transferToThirdCountries" : self.transferToThirdCountries,
            "dataPurposes" : self.dataPurposes,
            "dataPurposesInfo" : self.dataPurposesInfo,
            "dataRecipientsList" : self.dataRecipientsList,
            "descriptionOfService" : self.descriptionOfService,
            "history" : self.history,
            "historyDescription" : self.historyDescription,
            "legalBasisList" : self.legalBasisList,
            "legalBasisInfo" : self.legalBasisInfo,
            "processingCompanyTitle" : self.processingCompanyTitle,
            "technologiesUsed" : self.technologiesUsed,
            "technologiesUsedInfo" : self.technologiesUsedInfo,
            "cookiePolicyInfo" : self.cookiePolicyInfo,
            "optOut" : self.optOut,
            "policyOf" : self.policyOf,
            "imprintLinkText" : self.imprintLinkText,
            "privacyPolicyLinkText" : self.privacyPolicyLinkText,
            "categories" : self.categories,
            "anyDomain" : self.anyDomain,
            "day" : self.day,
            "days" : self.days,
            "domain" : self.domain,
            "duration" : self.duration,
            "informationLoadingNotPossible" : self.informationLoadingNotPossible,
            "hour" : self.hour,
            "hours" : self.hours,
            "identifier" : self.identifier,
            "maximumAgeCookieStorage" : self.maximumAgeCookieStorage,
            "minute" : self.minute,
            "minutes" : self.minutes,
            "month" : self.month,
            "months" : self.months,
            "multipleDomains" : self.multipleDomains,
            "no" : self.no,
            "nonCookieStorage" : self.nonCookieStorage,
            "seconds" : self.seconds,
            "session" : self.session,
            "loadingStorageInformation" : self.loadingStorageInformation,
            "storageInformation" : self.storageInformation,
            "detailedStorageInformation" : self.detailedStorageInformation,
            "tryAgain" : self.tryAgain,
            "type" : self.type,
            "year" : self.year,
            "years" : self.years,
            "yes" : self.yes,
            "storageInformationDescription" : self.storageInformationDescription,
            "btnBannerReadMore" : self.btnBannerReadMore,
            "linkToDpaInfo" : self.linkToDpaInfo,
            "second" : self.second,
            "consent" : self.consent,
            "secondLayerDescriptionHtml" : self.secondLayerDescriptionHtml,
            "secondLayerTitle" : self.secondLayerTitle,
            "btnMore": self.btnMore,
            "retentionPeriod": self.retentionPeriod,
            "explicit": self.explicit_ as Any,
            "transferToThirdCountriesInfo": self.transferToThirdCountriesInfo,
            "more": self.more,
            "headerModal": self.headerModal,
            "furtherInformationOptOut": self.furtherInformationOptOut,
            "cookiePolicyLinkText": self.cookiePolicyLinkText,
            "noImplicit": self.noImplicit,
            "yesImplicit": self.yesImplicit,
            "readLess": self.readLess
        ]
    }
}

extension CCPASettings {
    func serialize() -> Any {
        return [
            "optOutNoticeLabel" : self.optOutNoticeLabel,
            "btnSave" : self.btnSave,
            "firstLayerTitle" : self.firstLayerTitle,
            "isActive" : self.isActive,
            "showOnPageLoad" : self.showOnPageLoad,
            "reshowAfterDays" : self.reshowAfterDays,
            "iabAgreementExists" : self.iabAgreementExists,
            "appFirstLayerDescription" : self.appFirstLayerDescription as Any,
            "firstLayerMobileDescriptionIsActive" : self.firstLayerMobileDescriptionIsActive,
            "firstLayerMobileDescription" : self.firstLayerMobileDescription as Any,
            "secondLayerTitle" : self.secondLayerTitle as Any,
            "secondLayerDescription" : self.secondLayerDescription as Any,
            "secondLayerHideLanguageSwitch" : self.secondLayerHideLanguageSwitch,
            "btnMoreInfo" : self.btnMoreInfo as Any,
        ]
    }
}


extension TCF2Settings {
    func serialize() -> Any {
        return [
            "firstLayerTitle" : self.firstLayerTitle,
            "secondLayerTitle" : self.secondLayerTitle,
            "tabsPurposeLabel" : self.tabsPurposeLabel,
            "tabsVendorsLabel" : self.tabsVendorsLabel,
            "labelsFeatures" : self.labelsFeatures,
            "labelsIabVendors" : self.labelsIabVendors,
            "labelsNonIabPurposes" : self.labelsNonIabPurposes,
            "labelsNonIabVendors" : self.labelsNonIabVendors,
            "labelsPurposes" : self.labelsPurposes,
            "vendorFeatures" : self.vendorFeatures,
            "vendorLegitimateInterestPurposes" : self.vendorLegitimateInterestPurposes,
            "vendorPurpose" : self.vendorPurpose,
            "vendorSpecialFeatures" : self.vendorSpecialFeatures,
            "vendorSpecialPurposes" : self.vendorSpecialPurposes,
            "togglesConsentToggleLabel" : self.togglesConsentToggleLabel,
            "togglesLegIntToggleLabel" : self.togglesLegIntToggleLabel,
            "buttonsAcceptAllLabel" : self.buttonsAcceptAllLabel,
            "buttonsDenyAllLabel" : self.buttonsDenyAllLabel,
            "buttonsSaveLabel" : self.buttonsSaveLabel,
            "linksManageSettingsLabel" : self.linksManageSettingsLabel,
            "linksVendorListLinkLabel" : self.linksVendorListLinkLabel,
            "cmpId" : self.cmpId,
            "cmpVersion" : self.cmpVersion,
            "categoriesOfDataLabel" : self.categoriesOfDataLabel,
            "dataRetentionPeriodLabel" : self.dataRetentionPeriodLabel,
            "legitimateInterestLabel" : self.legitimateInterestLabel,
            "version" : self.version,
            "examplesLabel" : self.examplesLabel,
            "firstLayerHideToggles" : self.firstLayerHideToggles,
            "secondLayerHideToggles" : self.secondLayerHideToggles,
            "hideLegitimateInterestToggles" : self.hideLegitimateInterestToggles,
            "firstLayerHideButtonDeny" : Bool(from: self.firstLayerHideButtonDeny) as Any,
            "secondLayerHideButtonDeny" : self.secondLayerHideButtonDeny,
            "publisherCountryCode" : self.publisherCountryCode,
            "purposeOneTreatment" : self.purposeOneTreatment,
            "selectedVendorIds" : self.selectedVendorIds,
            "gdprApplies" : self.gdprApplies,
            "selectedStacks" : self.selectedStacks,
            "disabledSpecialFeatures" : self.disabledSpecialFeatures,
            "firstLayerShowDescriptions" : self.firstLayerShowDescriptions,
            "hideNonIabOnFirstLayer" : self.hideNonIabOnFirstLayer,
            "resurfacePeriodEnded" : self.resurfacePeriodEnded,
            "resurfacePurposeChanged" : self.resurfacePurposeChanged,
            "resurfaceVendorAdded" : self.resurfaceVendorAdded,
            "firstLayerDescription" : self.firstLayerDescription as Any,
            "firstLayerAdditionalInfo" : self.firstLayerAdditionalInfo as Any,
            "secondLayerDescription" : self.secondLayerDescription as Any,
            "togglesSpecialFeaturesToggleOn" : self.togglesSpecialFeaturesToggleOn as Any,
            "togglesSpecialFeaturesToggleOff" : self.togglesSpecialFeaturesToggleOff as Any,
            "appLayerNoteResurface" : self.appLayerNoteResurface as Any,
            "firstLayerNoteResurface" : self.firstLayerNoteResurface as Any,
            "firstLayerMobileVariant": self.firstLayerMobileVariant?.name ?? "",
            "showDataSharedOutsideEUText": self.showDataSharedOutsideEUText,
            "dataSharedOutsideEUText": self.dataSharedOutsideEUText as Any,
            "vendorIdsOutsideEUList": self.vendorIdsOutsideEUList,
            "scope": self.scope.name,
            "changedPurposes": self.changedPurposes?.serialize() as Any,
            "acmV2Enabled" : self.acmV2Enabled,
            "selectedATPIds" : self.selectedATPIds
        ]
    }
}


extension UsercentricsCustomization {
    func serialize() -> Any {
        return [
            "color" : self.color?.serialize(),
            "font" : self.font?.serialize(),
            "logoUrl" : self.logoUrl,
            "borderRadiusLayer" : self.borderRadiusLayer,
            "borderRadiusButton" : self.borderRadiusButton,
            "overlayOpacity" : self.overlayOpacity,
        ]
    }
}



extension CustomizationFont {
    func serialize() -> Any {
        return [
            "family" : self.family as Any,
            "size" : self.size as Any,
        ]
    }
}



extension CustomizationColor {
    func serialize() -> Any {
        return [
            "primary" : self.primary,
            "acceptBtnText" : self.acceptBtnText,
            "acceptBtnBackground" : self.acceptBtnBackground,
            "denyBtnText" : self.denyBtnText,
            "denyBtnBackground" : self.denyBtnBackground,
            "saveBtnText" : self.saveBtnText,
            "saveBtnBackground" : self.saveBtnBackground,
            "linkIcon" : self.linkIcon,
            "linkFont" : self.linkFont,
            "text" : self.text,
            "layerBackground" : self.layerBackground,
            "overlay" : self.overlay,
            "toggleInactiveBackground" : self.toggleInactiveBackground,
            "toggleInactiveIcon" : self.toggleInactiveIcon,
            "toggleActiveBackground" : self.toggleActiveBackground,
            "toggleActiveIcon" : self.toggleActiveIcon,
            "toggleDisabledBackground" : self.toggleDisabledBackground,
            "toggleDisabledIcon" : self.toggleDisabledIcon,
            "secondLayerTab" : self.secondLayerTab,
        ]
    }
}



extension FirstLayer {
    func serialize() -> Any {
        return [
            "hideButtonDeny" : Bool(from: self.hideButtonDeny) as Any,
        ]
    }
}



extension SecondLayer {
    func serialize() -> Any {
        return [
            "tabsCategoriesLabel" : self.tabsCategoriesLabel,
            "tabsServicesLabel" : self.tabsServicesLabel,
            "acceptButtonText" : self.acceptButtonText as Any,
            "denyButtonText" : self.denyButtonText as Any,
            "hideButtonDeny" : Bool(from: self.hideButtonDeny) as Any,
            "hideLanguageSwitch" : Bool(from: self.hideLanguageSwitch) as Any,
            "hideTogglesForServices": self.hideTogglesForServices,
            "hideDataProcessingServices": self.hideDataProcessingServices,
        ]
    }
}



extension UsercentricsService {
    func serialize() -> Any {
        return [
            "templateId" : self.templateId as Any,
            "version" : self.version as Any,
            "type" : self.type as Any,
            "dataProcessor" : self.dataProcessor as Any,
            "dataPurposes" : self.dataPurposes,
            "processingCompany" : self.processingCompany as Any,
            "nameOfProcessingCompany" : self.nameOfProcessingCompany,
            "addressOfProcessingCompany" : self.addressOfProcessingCompany,
            "descriptionOfService" : self.descriptionOfService,
            "languagesAvailable" : self.languagesAvailable,
            "dataCollectedList" : self.dataCollectedList,
            "dataPurposesList" : self.dataPurposesList,
            "dataRecipientsList" : self.dataRecipientsList,
            "legalBasisList" : self.legalBasisList,
            "retentionPeriodList" : self.retentionPeriodList,
            "subConsents" : self.subConsents as Any,
            "language" : self.language,
            "linkToDpa" : self.linkToDpa,
            "legalGround" : self.legalGround,
            "optOutUrl" : self.optOutUrl,
            "policyOfProcessorUrl" : self.policyOfProcessorUrl,
            "categorySlug" : self.categorySlug as Any,
            "retentionPeriodDescription" : self.retentionPeriodDescription,
            "dataProtectionOfficer" : self.dataProtectionOfficer,
            "privacyPolicyURL" : self.privacyPolicyURL,
            "cookiePolicyURL" : self.cookiePolicyURL,
            "locationOfProcessing" : self.locationOfProcessing,
            "dataCollectedDescription" : self.dataCollectedDescription as Any,
            "thirdCountryTransfer" : self.thirdCountryTransfer,
            "description" : self.description_ as Any,
            "cookieMaxAgeSeconds" : self.cookieMaxAgeSeconds as Any,
            "usesNonCookieAccess" : Bool(from: self.usesNonCookieAccess) as Any,
            "deviceStorageDisclosureUrl" : self.deviceStorageDisclosureUrl as Any,
            "isDeactivated" : Bool(from: self.isDeactivated) as Any,
            "disableLegalBasis" : Bool(from: self.disableLegalBasis) as Any,
            "isEssential" : self.isEssential,
            "technologyUsed": self.technologyUsed as Any,
            "deviceStorage": self.deviceStorage.serialize(),
            "isHidden": self.isHidden
        ]
    }
}

extension UsercentricsCategory {
    func serialize() -> Any? {
        return [
            "categorySlug" : self.categorySlug,
            "label" : self.label as Any,
            "description" : self.description_ as Any,
            "isEssential" : self.isEssential
        ]
    }
}

extension UsercentricsLocation {
    func serialize() -> Any {
        return [
            "countryCode" : self.countryCode,
            "regionCode" : self.regionCode,
            "isInEU" : self.isInEU(),
            "isInUS" : self.isInUS(),
            "isInCalifornia" : self.isInCalifornia(),
        ] as [String : Any]
    }
}

extension VariantsSettings {
    func serialize() -> Any {
        return [
            "enabled": enabled,
            "experimentsJson": experimentsJson,
            "activateWith": activateWith,
        ] as [String : Any]
    }
}

extension PublishedApp {
    func serialize() -> Any {
        return [
            "bundleId": bundleId,
            "platform": platform.name,
        ]
    }
}

extension TCF2ChangedPurposes {
    func serialize() -> Any {
        return [
            "purposes": purposes,
            "legIntPurposes": legIntPurposes
        ]
    }
}


extension LegalBasisLocalization {
    func serialize() -> Any {
        return [
            "labelsAria": self.labelsAria?.serialize(),
            "data": self.data
        ]
    }
}

extension TranslationAriaLabels {
    func serialize() -> Any {
        return [
            "acceptAllButton": self.acceptAllButton,
            "ccpaButton": self.ccpaButton,
            "ccpaMoreInformation": self.ccpaMoreInformation,
            "closeButton": self.closeButton,
            "collapse": self.collapse,
            "cookiePolicyButton": self.cookiePolicyButton,
            "copyControllerId": self.copyControllerId,
            "denyAllButton": self.denyAllButton,
            "expand": self.expand,
            "fullscreenButton": self.fullscreenButton,
            "imprintButton": self.imprintButton,
            "languageSelector": self.languageSelector,
            "privacyButton": self.privacyButton,
            "privacyPolicyButton": self.privacyPolicyButton,
            "saveButton": self.saveButton,
            "serviceInCategoryDetails": self.serviceInCategoryDetails,
            "servicesInCategory": self.servicesInCategory,
            "tabButton": self.tabButton,
            "usercentricsCMPButtons": self.usercentricsCMPButtons,
            "usercentricsCMPContent": self.usercentricsCMPContent,
            "usercentricsCMPHeader": self.usercentricsCMPHeader,
            "usercentricsCMPUI": self.usercentricsCMPUI,
            "usercentricsCard": self.usercentricsCard,
            "usercentricsList": self.usercentricsList,
            "vendorConsentToggle": self.vendorConsentToggle,
            "vendorDetailedStorageInformation": self.vendorDetailedStorageInformation,
            "vendorLegIntToggle": self.vendorLegIntToggle
        ]
    }
}
